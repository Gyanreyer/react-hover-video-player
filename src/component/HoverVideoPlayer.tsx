import React, { useRef, useImperativeHandle, useEffect, useState } from 'react';

import useSetAdditionalAttributesOnVideo from '../hooks/useSetAdditionalAttributesOnVideo';
import useFormatVideoSrc from '../hooks/useFormatVideoSrc';
import useFormatVideoCaptions from '../hooks/useFormatVideoCaptions';
import useHoverTargetElement from '../hooks/useHoverTargetElement';
import useManageHoverEvents from '../hooks/useManageHoverEvents';

import {
  isVideoElementPaused,
  isVideoElementPlaying,
} from '../utils/videoElementPlaybackStates';

import {
  expandToFillContainerStyle,
  containerSizingStyles,
  pausedOverlayWrapperSizingStyles,
  videoSizingStyles,
} from './HoverVideoPlayer.styles';
import { HoverVideoPlayerProps, VideoSource } from '../HoverVideoPlayer.types';

/**
 * @component HoverVideoPlayer
 * @license MIT
 *
 * @param {HoverVideoPlayerProps} props
 */
export default function HoverVideoPlayer({
  videoSrc,
  videoCaptions = null,
  focused = false,
  disableDefaultEventHandling = false,
  hoverTarget = null,
  onHoverStart = null,
  onHoverEnd = null,
  hoverOverlay = null,
  pausedOverlay = null,
  loadingOverlay = null,
  loadingStateTimeout = 200,
  overlayTransitionDuration = 400,
  playbackStartDelay = 0,
  restartOnPaused = false,
  unloadVideoOnPaused = false,
  playbackRangeStart = null,
  playbackRangeEnd = null,
  muted = true,
  volume = 1,
  loop = true,
  preload = null,
  crossOrigin = null,
  controls = false,
  controlsList = null,
  disableRemotePlayback = true,
  disablePictureInPicture = true,
  style = null,
  hoverOverlayWrapperClassName = null,
  hoverOverlayWrapperStyle = null,
  pausedOverlayWrapperClassName = null,
  pausedOverlayWrapperStyle = null,
  loadingOverlayWrapperClassName = null,
  loadingOverlayWrapperStyle = null,
  videoId = null,
  videoClassName = null,
  videoRef: forwardedVideoRef = null,
  videoStyle = null,
  sizingMode = 'video',
  shouldSuppressPlaybackInterruptedErrors = true,
  ...spreadableProps
}: HoverVideoPlayerProps): JSX.Element {
  // Element refs
  const containerRef = useRef(null);
  const videoRef = useRef<HTMLVideoElement>(null);
  // Forward out local videoRef along to the videoRef prop
  useImperativeHandle(forwardedVideoRef, () => videoRef.current);

  // Effect sets attributes on the video which can't be done via props
  useSetAdditionalAttributesOnVideo(
    videoRef,
    muted,
    volume,
    disableRemotePlayback,
    disablePictureInPicture
  );

  // Get the hover target element from the hoverTarget prop, or default to the component's container div
  const hoverTargetElement = useHoverTargetElement(hoverTarget || containerRef);

  // Keep a ref for the time which the video should be started from next time it is played
  // This is useful if the video gets unloaded and we want to restore it to the time it was
  // at before if the user tries playing it again
  const nextVideoStartTimeRef = useRef(null);

  // Parse the sources and captions into formatted arrays that we can use to
  // render <source> and <track> elements for the video
  const formattedVideoCaptions = useFormatVideoCaptions(videoCaptions);
  const formattedVideoSources = useFormatVideoSrc(
    videoSrc,
    playbackRangeStart,
    playbackRangeEnd
  );
  // Keep a ref to the previous formatted video sources so we can track when the video sources change
  const previousFormattedVideoSourcesRef = useRef<VideoSource[]>(
    formattedVideoSources
  );

  /**
   * Attempts to load and play the video.
   * Storing this on a ref because we don't really want to worry about triggering re-renders when
   * any of this function's dependencies change; it should only be called when the
   * player is hovered/focused.
   */
  const attemptToPlayVideoRef = useRef<() => void>();
  attemptToPlayVideoRef.current = () => {
    const videoElement = videoRef.current;

    if (nextVideoStartTimeRef.current !== null) {
      videoElement.currentTime = nextVideoStartTimeRef.current;
    }

    videoElement.play().catch((error: DOMException) => {
      // If shouldSuppressPlaybackInterruptedErrors is true and this is an AbortError, do nothing instead of logging it.
      if (
        shouldSuppressPlaybackInterruptedErrors &&
        error.name === 'AbortError'
      ) {
        return;
      }

      // Additional handling for when browsers block playback for unmuted videos.
      // This is unfortunately necessary because most modern browsers do not allow playing videos with audio
      //  until the user has "interacted" with the page by clicking somewhere at least once; mouseenter events
      //  don't count.
      // If the video isn't muted and playback failed with a `NotAllowedError`, this means the browser blocked
      // playing the video because the user hasn't clicked anywhere on the page yet.
      if (!videoElement.muted && error.name === 'NotAllowedError') {
        console.warn(
          'HoverVideoPlayer: Playback with sound was blocked by the browser. Attempting to play again with the video muted; audio will be restored if the user clicks on the page.'
        );
        // Mute the video and attempt to play again
        videoElement.muted = true;
        videoElement.play();

        // When the user clicks on the document, unmute the video since we should now
        // be free to play audio
        const onClickDocument = () => {
          videoElement.muted = false;

          // Clean up the event listener so it is only fired once
          document.removeEventListener('click', onClickDocument);
        };
        document.addEventListener('click', onClickDocument);
      } else {
        // Log any other playback errors with console.error
        console.error(`HoverVideoPlayer: ${error.message}`);
      }
    });
  };

  /**
   * Attempts to pause the video.
   * Storing this on a ref because we don't really want to worry about triggering re-renders when
   * any of this function's dependencies change; it should only be called when the
   * player is un-hovered/focused.
   */
  const attemptToPauseVideoRef = useRef<() => void>();
  attemptToPauseVideoRef.current = () => {
    const videoElement = videoRef.current;

    videoElement.pause();

    // Performing post-save cleanup tasks in here rather than the onPause listener
    // because onPause can also be called when the video reaches the end of a playback range
    // and it's just simpler to deal with that separately
    if (restartOnPaused) {
      videoElement.currentTime = playbackRangeStart || 0;
    }
    nextVideoStartTimeRef.current = videoElement.currentTime;

    setIsPlaying(false);
  };

  const [isPlaying, setIsPlaying] = useState(false);
  const [isHovering, setIsHovering] = useState(false);

  useEffect(() => {
    if (!isPlaying) {
      // When the video isn't playing, check if the sources loaded in the video
      // have changed from what the video currently has loaded. If so, we'll call videoElement.load()
      // to trigger a reload with the new source
      const shouldReloadVideoSources =
        previousFormattedVideoSourcesRef.current !== formattedVideoSources;

      if (shouldReloadVideoSources) {
        previousFormattedVideoSourcesRef.current = formattedVideoSources;

        const videoElement = videoRef.current;
        // If the video element doesn't have a loaded source or the source has changed since the
        // last time we played the video, make sure to force the video to load the most up-to-date sources
        videoElement.load();
        // Reset the next start time to the start of the video
        nextVideoStartTimeRef.current = playbackRangeStart || 0;
      }
    }
  }, [formattedVideoSources, isPlaying, playbackRangeStart]);

  const hasPausedOverlay = Boolean(pausedOverlay);
  const hasHoverOverlay = Boolean(hoverOverlay);

  // If we have a paused or hover overlay, the player should wait
  // for the overlay(s) to finish transitioning back in before we
  // pause the video
  const shouldWaitForOverlayTransitionBeforePausing =
    hasPausedOverlay || hasHoverOverlay;

  useEffect(() => {
    if (!hoverTargetElement) return undefined;

    let playbackStartTimeout: number | null = null;
    let pauseTimeout: number | null = null;

    const cancelTimeouts = () => {
      // Cancel any previously active pause or playback attempts
      window.clearTimeout(playbackStartTimeout);
      window.clearTimeout(pauseTimeout);
    };

    const attemptToPlayVideoOnHover = () => {
      cancelTimeouts();

      setIsHovering(true);

      // We only need to attempt to play if the video is currently paused
      if (isVideoElementPaused(videoRef.current)) {
        if (playbackStartDelay) {
          playbackStartTimeout = window.setTimeout(
            () => attemptToPlayVideoRef.current(),
            playbackStartDelay
          );
        } else {
          attemptToPlayVideoRef.current();
        }
      }
    };
    const attemptToPauseVideoOnHoverEnd = () => {
      cancelTimeouts();

      setIsHovering(false);

      if (
        // We only need to delay a pause attempt if the video is currently playing
        isVideoElementPlaying(videoRef.current) &&
        shouldWaitForOverlayTransitionBeforePausing
      ) {
        pauseTimeout = window.setTimeout(
          () => attemptToPauseVideoRef.current(),
          overlayTransitionDuration
        );
      } else {
        attemptToPauseVideoRef.current();
      }
    };

    hoverTargetElement.addEventListener(
      'hvp:hoverStart',
      attemptToPlayVideoOnHover
    );
    hoverTargetElement.addEventListener(
      'hvp:hoverEnd',
      attemptToPauseVideoOnHoverEnd
    );

    return () => {
      hoverTargetElement.removeEventListener(
        'hvp:hoverStart',
        attemptToPlayVideoOnHover
      );
      hoverTargetElement.removeEventListener(
        'hvp:hoverEnd',
        attemptToPauseVideoOnHoverEnd
      );
      cancelTimeouts();
    };
  }, [
    hoverTargetElement,
    overlayTransitionDuration,
    playbackStartDelay,
    shouldWaitForOverlayTransitionBeforePausing,
  ]);

  useManageHoverEvents(
    hoverTargetElement,
    focused,
    disableDefaultEventHandling,
    onHoverStart,
    onHoverEnd
  );

  // We should attempt to play the video if the user is hovering over it or the `focused` override prop is enabled
  // const shouldPlayVideo = isHoveringOverVideo || focused;

  const hasLoadingOverlay = Boolean(loadingOverlay);

  // If the video's sources should be unloaded when it's paused and the video is not currently active, we can unload the video's sources.
  // We will remove the video's <source> tags in this render and then call video.load() in an effect to
  // fully unload the video
  const shouldUnloadVideo = unloadVideoOnPaused && !isHovering && !isPlaying;

  useEffect(() => {
    // If shouldUnloadVideo is true, this effect is being run after the video's
    // sources have been removed, so call load on the video to unload any sources it currently has loaded
    if (shouldUnloadVideo) {
      const videoElement = videoRef.current;
      videoElement.load();
    }
  }, [shouldUnloadVideo]);

  const shouldShowLoadingOverlay = isHovering && !isPlaying;
  // Show a paused overlay when the user isn't hovering or when the user is hovering
  // but the video is still loading
  const shouldShowPausedOverlay = !isHovering || (isHovering && !isPlaying);

  const isUsingPlaybackRange =
    playbackRangeStart !== null || playbackRangeEnd !== null;

  return (
    <div
      data-testid="hover-video-player-container"
      ref={containerRef}
      style={{
        ...containerSizingStyles[sizingMode],
        position: 'relative',
        ...style,
      }}
      {...spreadableProps}
    >
      {hasPausedOverlay ? (
        <div
          style={{
            ...pausedOverlayWrapperSizingStyles[sizingMode],
            zIndex: 1,
            opacity: shouldShowPausedOverlay ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms`,
            // Disable pointer events on the paused overlay when it's hidden
            pointerEvents: shouldShowPausedOverlay ? 'auto' : 'none',
            ...pausedOverlayWrapperStyle,
          }}
          className={pausedOverlayWrapperClassName}
          data-testid="paused-overlay-wrapper"
        >
          {pausedOverlay}
        </div>
      ) : null}
      {hasLoadingOverlay ? (
        <div
          style={{
            ...expandToFillContainerStyle,
            zIndex: 2,
            opacity: shouldShowLoadingOverlay ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms ${
              shouldShowLoadingOverlay ? loadingStateTimeout : 0
            }ms`,
            // Disable pointer events on the loading overlay when it's hidden
            pointerEvents: shouldShowLoadingOverlay ? 'auto' : 'none',
            ...loadingOverlayWrapperStyle,
          }}
          className={loadingOverlayWrapperClassName}
          data-testid="loading-overlay-wrapper"
        >
          {loadingOverlay}
        </div>
      ) : null}
      {hasHoverOverlay ? (
        <div
          style={{
            ...expandToFillContainerStyle,
            zIndex: 3,
            // Show the hover overlay when the player is hovered/playing
            opacity: isHovering ? 1 : 0,
            transition: `opacity ${overlayTransitionDuration}ms`,
            // Disable pointer events on the hover overlay when it's hidden
            pointerEvents: isHovering ? 'auto' : 'none',
            ...hoverOverlayWrapperStyle,
          }}
          className={hoverOverlayWrapperClassName}
          data-testid="hover-overlay-wrapper"
        >
          {hoverOverlay}
        </div>
      ) : null}
      {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
      <video
        // If a playback range is set, the loop attribute will not work correctly so there's no point in setting it here;
        // in that case, we will manually implement this behavior
        loop={isUsingPlaybackRange ? false : loop}
        playsInline
        preload={preload}
        crossOrigin={crossOrigin}
        ref={videoRef}
        style={{
          ...videoSizingStyles[sizingMode],
          objectFit: 'cover',
          ...videoStyle,
        }}
        controls={controls}
        controlsList={controlsList}
        className={videoClassName}
        id={videoId}
        data-testid="video-element"
        onPlaying={() => {
          setIsPlaying(true);
        }}
        onTimeUpdate={
          // If there's a playback range set, the traditional `loop` video prop won't work correctly so
          // we'll need watch the video's time as it plays and manually keep it within the bounds of the range
          isUsingPlaybackRange
            ? () => {
                const videoElement = videoRef.current;

                const maxVideoTime = playbackRangeEnd || videoElement.duration;
                const minVideoTime = playbackRangeStart || 0;

                const { currentTime } = videoElement;

                if (loop && currentTime >= maxVideoTime) {
                  // If the video should loop and is >= the max video time,
                  // loop it back around to the start
                  const startTime = playbackRangeStart || 0;
                  videoElement.currentTime = startTime;

                  // If the video is paused but the user is still hovering,
                  // meaning it should continue to play, call play() to keep it going
                  if (isHovering && isVideoElementPaused(videoElement)) {
                    videoElement.play();
                  }
                } else if (currentTime > maxVideoTime) {
                  // If the video shouldn't loop but we've exceeded the max video time,
                  // clamp it to the max time and pause it
                  videoElement.pause();
                  videoElement.currentTime = maxVideoTime;
                } else if (currentTime < minVideoTime) {
                  // If the video's time somehow ended up before the min video time,
                  // clamp it to the min time
                  videoElement.currentTime = minVideoTime;
                }
              }
            : null
        }
      >
        {!shouldUnloadVideo &&
          // Only render sources for the video if it is not unloaded
          formattedVideoSources.map(({ src, type }) => (
            <source key={src} src={src} type={type} />
          ))}
        {formattedVideoCaptions.map(
          ({ src, srcLang, label, kind, default: isDefault }) => (
            <track
              key={src}
              kind={kind}
              src={src}
              srcLang={srcLang}
              label={label}
              default={isDefault}
            />
          )
        )}
      </video>
    </div>
  );
}
