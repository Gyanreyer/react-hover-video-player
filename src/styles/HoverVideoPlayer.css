.Container {
  position: relative;
}

.PausedOverlayContainer {
  position: relative;
  display: block;
  width: 100%;
  z-index: 1;
  pointer-events: none;
}

.LoadingOverlayContainer {
  z-index: 2;
  pointer-events: none;
}

.Video {
  width: 100%;
  display: block;
}

.ExpandToCoverContainer {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
}

.ShouldExpandToCoverVideo, .LoadingOverlayContainer, .ShouldExpandToFitOverlay {
  composes: ExpandToCoverContainer;
}

.PausedOverlayContainer > * {
  display: block;
}

.Video.ShouldExpandToFitOverlay {
  object-fit: cover;
}