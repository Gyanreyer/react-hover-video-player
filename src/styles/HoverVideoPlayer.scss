.Container {
  position: relative;
}

.PausedOverlayContainer, .LoadingOverlayContainer {
  /* Overlay wrapper should completely cover the dimensions of the video */
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  pointer-events: none;
}

.LoadingOverlayContainer {
  z-index: 1;
}

.Video {
  width: 100%;
  display: block;
}