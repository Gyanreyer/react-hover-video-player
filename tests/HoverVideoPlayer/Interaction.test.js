import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import {
  expectVideoHasCorrectAttributes,
  addMockedFunctionsToVideoElement,
} from '../utils';
import HoverVideoPlayer from '../../src';

describe('Handles desktop mouse events correctly', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('mouseEnter event takes the video through its start flow correctly', async () => {
    const onStartingVideo = jest.fn();
    const onStartedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStartingVideo={onStartingVideo}
        onStartedVideo={onStartedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(getByTestId('hover-video-player-container'));

    // The onStartingVideo callback should have been called
    expect(onStartingVideo).toHaveBeenCalled();
    // The video's play function should have been called
    expect(videoElement.play).toHaveBeenCalled();

    // onStartedVideo shouldn't be called since the play promise is still pending
    expect(onStartedVideo).not.toHaveBeenCalled();

    // Wait until the play promise has resolved
    await act(() => playPromise);

    expect(onStartedVideo).toHaveBeenCalled();
  });

  test('mouseLeave event takes the video through its stop flow correctly', async () => {
    const onStoppingVideo = jest.fn();
    const onStoppedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStoppingVideo={onStoppingVideo}
        onStoppedVideo={onStoppedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    // Mouse out of the container to stop playing the video
    fireEvent.mouseLeave(playerContainer);

    // onStoppingVideo callback should have been called
    expect(onStoppingVideo).toHaveBeenCalled();
    // onStoppedVideo callback and video.pause should not be called until the timeout has completed
    expect(onStoppedVideo).not.toHaveBeenCalled();
    expect(videoElement.pause).not.toHaveBeenCalled();

    // Advance a sufficient amount of time for the pause timeout to have completed
    jest.advanceTimersByTime(500);

    expect(onStoppedVideo).toHaveBeenCalled();
    expect(videoElement.pause).toHaveBeenCalled();
  });
});

describe('Handles mobile touch events correctly', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('touchStart event takes the video through its start flow correctly', async () => {
    const onStartingVideo = jest.fn();
    const onStartedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStartingVideo={onStartingVideo}
        onStartedVideo={onStartedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    // Mouse over the container to start playing the video
    fireEvent.touchStart(getByTestId('hover-video-player-container'));

    // The onStartingVideo callback should have been called
    expect(onStartingVideo).toHaveBeenCalled();
    // The video's play function should have been called
    expect(videoElement.play).toHaveBeenCalled();

    // onStartedVideo shouldn't be called since the play promise is still pending
    expect(onStartedVideo).not.toHaveBeenCalled();

    // Wait until the play promise has resolved
    await act(() => playPromise);

    expect(onStartedVideo).toHaveBeenCalled();
  });

  test('touchStart events outside of the video takes it through its stop flow correctly', async () => {
    const onStoppingVideo = jest.fn();
    const onStoppedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStoppingVideo={onStoppingVideo}
        onStoppedVideo={onStoppedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    // Touch the container to start playing the video
    fireEvent.touchStart(playerContainer);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    // Touching an element inside of the player container should not start a stop attempt
    fireEvent.touchStart(videoElement);

    expect(onStoppingVideo).not.toHaveBeenCalled();
    expect(onStoppedVideo).not.toHaveBeenCalled();

    // Touching outside of the player container should start a stop attempt
    fireEvent.touchStart(document.body);

    // onStoppingVideo callback should have been called
    expect(onStoppingVideo).toHaveBeenCalled();
    // onStoppedVideo callback and video.pause should not be called until the timeout has completed
    expect(onStoppedVideo).not.toHaveBeenCalled();
    expect(videoElement.pause).not.toHaveBeenCalled();

    // Advance a sufficient amount of time for the pause timeout to have completed
    jest.advanceTimersByTime(500);

    expect(onStoppedVideo).toHaveBeenCalled();
    expect(videoElement.pause).toHaveBeenCalled();
  });
});

describe('Start and stop attempts interact with each other correctly', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('an attempt to start the video will correctly interrupt any attempts to stop it', async () => {
    const onStartingVideo = jest.fn();
    const onStartedVideo = jest.fn();
    const onStoppingVideo = jest.fn();
    const onStoppedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStartingVideo={onStartingVideo}
        onStartedVideo={onStartedVideo}
        onStoppingVideo={onStoppingVideo}
        onStoppedVideo={onStoppedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    expect(onStartingVideo).toHaveBeenCalledTimes(1);
    expect(onStartedVideo).toHaveBeenCalledTimes(0);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    expect(onStartedVideo).toHaveBeenCalledTimes(1);

    // We are now in playing state, so mouse out to stop again
    fireEvent.mouseLeave(playerContainer);

    // At this point the video should have begun its stop attempt but not completed it
    expect(onStoppingVideo).toHaveBeenCalledTimes(1);
    expect(onStoppedVideo).toHaveBeenCalledTimes(0);

    // Mouse back over to cancel the stop attempt and start a play attempt
    fireEvent.mouseEnter(playerContainer);

    // A second start attempt should have begun
    expect(onStartingVideo).toHaveBeenCalledTimes(2);
    expect(onStartedVideo).toHaveBeenCalledTimes(1);

    await act(() => playPromise);

    // onStartedVideo should not need to be called again since we prevented it from stopping
    expect(onStartedVideo).toHaveBeenCalledTimes(1);

    jest.advanceTimersByTime(500);

    // onStoppedVideo should still not have been called because it was cancelled
    expect(onStoppedVideo).toHaveBeenCalledTimes(0);
  });

  test('an attempt to stop the video will correctly reverse an active start attempt when it completes', async () => {
    const onStartingVideo = jest.fn();
    const onStartedVideo = jest.fn();
    const onStoppingVideo = jest.fn();
    const onStoppedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStartingVideo={onStartingVideo}
        onStartedVideo={onStartedVideo}
        onStoppingVideo={onStoppingVideo}
        onStoppedVideo={onStoppedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    expect(videoElement.paused).toBe(true);

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    expect(onStartingVideo).toHaveBeenCalledTimes(1);
    expect(onStartedVideo).toHaveBeenCalledTimes(0);
    expect(videoElement.paused).toBe(false);

    // Fire a mouseLeave event to kick off a stop attempt before the play promise has resolved
    fireEvent.mouseLeave(playerContainer);

    // At this point the video should have begun its stop attempt but not completed it
    expect(onStoppingVideo).toHaveBeenCalledTimes(1);
    expect(onStoppedVideo).toHaveBeenCalledTimes(0);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    // onStoppedVideo should have been called after the promise resolved and the video's `onPlaying` event fired
    expect(onStoppedVideo).toHaveBeenCalledTimes(1);

    // onStartedVideo shouldn't have been called because it was cancelled
    expect(onStartedVideo).toHaveBeenCalledTimes(0);

    expect(videoElement.paused).toBe(true);
  });

  test('an attempt to start the video when it is already playing will be ignored', async () => {
    const onStartingVideo = jest.fn();
    const onStartedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStartingVideo={onStartingVideo}
        onStartedVideo={onStartedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playPromise = addMockedFunctionsToVideoElement(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    // The start attempt should have succeededF
    expect(onStartingVideo).toHaveBeenCalledTimes(1);
    expect(onStartedVideo).toHaveBeenCalledTimes(1);

    // A second mouseEnter event should effectively be ignored
    fireEvent.mouseEnter(playerContainer);

    // Wait until the play promise has resolved
    await act(() => playPromise);

    // We should not have run through the start attempt flow again
    expect(onStartingVideo).toHaveBeenCalledTimes(1);
    expect(onStartedVideo).toHaveBeenCalledTimes(1);
  });

  test('an attempt to stop the video when it is already stopped will be ignored', () => {
    const onStoppingVideo = jest.fn();
    const onStoppedVideo = jest.fn();

    const { container, getByTestId } = render(
      <HoverVideoPlayer
        videoSrc="fake/video-file.mp4"
        onStoppingVideo={onStoppingVideo}
        onStoppedVideo={onStoppedVideo}
      />
    );

    expect(container).toMatchSnapshot();

    const videoElement = container.querySelector('video');
    expectVideoHasCorrectAttributes(videoElement);

    const playerContainer = getByTestId('hover-video-player-container');

    // Mouse out of the container even though it was never started properly
    fireEvent.mouseLeave(playerContainer);

    expect(onStoppingVideo).not.toHaveBeenCalled();
    expect(onStoppedVideo).not.toHaveBeenCalled();
  });
});

/**
 * TESTS THAT STILL NEED TO BE WRITTEN
 *
 * - shouldRestartOnVideoStopped
 * - video.play() throws an error
 * - isFocused prop
 */
