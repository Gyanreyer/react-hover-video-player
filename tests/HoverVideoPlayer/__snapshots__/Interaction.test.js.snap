// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handles desktop mouse events correctly mouseEnter event takes the video through its start flow correctly 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles desktop mouse events correctly mouseLeave event takes the video through its stop flow correctly 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles mobile touch events correctly touchStart event takes the video through its start flow correctly 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles mobile touch events correctly touchStart events outside of the video takes it through its stop flow correctly 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Start and stop attempts interact with each other correctly an attempt to start the video when it is already playing will be ignored 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Start and stop attempts interact with each other correctly an attempt to start the video will correctly interrupt any attempts to stop it 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Start and stop attempts interact with each other correctly an attempt to stop the video when it is already stopped will be ignored 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Start and stop attempts interact with each other correctly an attempt to stop the video will correctly reverse an active start attempt when it completes 1`] = `
.emotion-1 {
  position: relative;
}

.emotion-0 {
  width: 100%;
  display: block;
  object-fit: cover;
}

<div>
  <div
    class="emotion-1"
    data-testid="hover-video-player-container"
  >
    <video
      class="emotion-0"
      loop=""
      playsinline=""
      preload="metadata"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;
